### ISSUE TITLE
SimpleHTTPServer import causes stack overflow

### ISSUE THREAD

**Issue by slide** [IronLanguages/ironpython2#40]
_From @jpmaterial on April 19, 2017 12:52_

The simple act of importing SimpleHTTPServer causes a stack overflow. I'm on Windows 10. It's the same under cmd.exe, PowerShell and MinGW. This is my complete console output:

<pre>IronPython 2.7.7 (2.7.7.0) on .NET 4.0.30319.42000 (64-bit)
Type "help", "copyright", "credits" or "license" for more information.
>>> import BaseHTTPServer
>>> import SimpleHTTPServer

Process is terminated due to StackOverflowException.</pre>

_Copied from original issue: IronLanguages/main#1628_

**Comment by slide** [comment]
I just tried this on my codebase that has this fix in it #1625 and I don't see the StackOverflowException.

**Comment by slide** [comment]
_From @jpmaterial on April 19, 2017 15:3_

That's good. When will the next release with this fix be?

**Comment by slide** [comment]
I'd like to do one within the next couple of weeks.

**Comment by slide** [comment]
_From @slozier on April 19, 2017 17:23_

@slide I don't think this one is related to #1625. I can reproduce over here on 64-bit (although only intermittently).

**Comment by slide** [comment]
Ok, we can continue to look into it of course.

**Comment by slide** [comment]
_From @slozier on April 20, 2017 1:56_

May be related to #1070.

**Comment by slide** [comment]
_From @slozier on April 20, 2017 2:16_

Did a bit of digging and boiled it down to the following repro:
```Python
def test_gh1628():
    def error():
        raise Exception()

    for _ in xrange(100000):
        try:
            error()
        except Exception:
            continue
```
This seems to fail somewhat consistently for me when run with ipy64.

**Comment by slide** [comment]
_From @jpmaterial on April 20, 2017 6:53_

Bingo, that reproduction produces a stack overflow for me too.

**Comment by slide** [comment]
_From @slozier on May 17, 2017 1:31_

I think this is the same issue as #1570

**Comment by slozier** [comment]
https://github.com/IronLanguages/main/pull/1167 provides a potential fix for this issue.

**Comment by lostmsu** [comment]
@slozier this seems to be critical :(

**Comment by canol** [comment]
Has there been a fix for this issue or any workarounds that I can apply?

**Comment by slozier** [comment]
@canol While not a fix for the underlying issue, the code which causes this particular failure can be rewritten to (hopefully) avoid the StackOverflowException. Here's the patched file:
[mimetypes.zip](https://github.com/IronLanguages/ironpython2/files/1270724/mimetypes.zip)

**Comment by canol** [comment]
Thank you for the patch. I replaced my mimetypes.py in "C:\Program Files (x86)\IronPython 2.7\Lib" with the file you have attached but it still throws the StackOverflowException exception when I execute "ipy -X:Frames -m ensurepip". I looked if there is a .pyc file in the same folder but couldn't find.

**Comment by canol** [comment]
I put some print statements in between lines to find out which line throws the exception. Looks like the line:

`    pip.main(args)`

in "IronPython 2.7\Lib\ensurepip\__init__.py" at line 48 throws the exception which is called after the edits you've done to mimetypes.py

**Comment by slozier** [comment]
Just to clear things it, was your initial StackOverflowException with ensurepip or with SimpleHTTPServer? The patch I provided is only a work around the SimpleHTTPServer issue. If the issue was with ensurepip, then it's possible it's the same underlying issue (although I've never run into it with ensurepip), or it's a completely different StackOverflowException issue...

**Comment by canol** [comment]
Sorry about the confusion, yes I am actually trying to install pip by executing `ipy -X:Frames -m ensurepip` command and all the errors I had was from executing this command. I came here from the issue #42

**Comment by slozier** [comment]
Thanks for the followup. I'm surprised the patch didn't work since the other thread mentions the ensurepip  failure being in mimetypes.. Will have to look into it some more. Have you tried `ipy.exe -X:Frames -X:Debug -m ensurepip`?

**Comment by slide** [comment]
Is the mimetypes.py being used inside the ensurepip packaged wheel files?

**Comment by canol** [comment]
I tried it now, the output is the same: "Process is terminated due to StackOverflowException." and it is encountered again at the same line.

But this time it takes quite a while before it prints that statement. Does it write logs somewhere that I can post here?

**Comment by slozier** [comment]
I'm not sure about the logging...

The patched mimetypes.py file makes the problem go away over here. Assuming mimetypes is causing the issue, a mode drastic "workaround" you could try would be to modify the `read_windows_registry` method in mimetypes.py and just make it return directly.

**Comment by canol** [comment]
FYI, I was having this issue on previous release but I just tried 2.7.8 Beta1 and this issue is not present. I can do `import SimpleHTTPServer` without an error and I can install pip by executing `ipy.exe -X:Frames -X:Debug -m ensurepip` successfully.

**Comment by slide** [comment]
FYI, the -X:Frames is no longer required, frames are enabled by default starting from 2.7.8

**Comment by slide** [comment]
Looks like this issue was already resolved in 2.7.8

**Comment by slozier** [comment]
Note: I don't think the underlying issue is resolved but it should occur less frequently.
